internal/setup/setup_test.go

```diff
@@ -1101,3 +1149,66 @@ func runParseReleaseTests(c *C, tests []setupTest) {
 		}
 	}
 }
+
+var sliceKeyTests = []struct {
+	input    string
+	expected setup.SliceKey
+	err      string
+}{{
+	input:    "foo_bar",
+	expected: setup.SliceKey{Package: "foo", Slice: "bar"},
+}, {
+	input:    "fo_bar",
+	expected: setup.SliceKey{Package: "fo", Slice: "bar"},
+}, {
+	input: "foo_ba",
+	err:   `invalid slice reference: "foo_ba"`,
+}, {
+	input: "f_bar",
+	err:   `invalid slice reference: "f_bar"`,
+}, {
+	input:    "1234_bar",
+	expected: setup.SliceKey{Package: "1234", Slice: "bar"},
+}, {
+	input: "1234_789",
+	err:   `invalid slice reference: "1234_789"`,
+}, {
+	input:    "foo1.1-2-3_bar",
+	expected: setup.SliceKey{Package: "foo1.1-2-3", Slice: "bar"},
+}, {
+	input: "chicken_bar.b.q",
+	err:   `invalid slice reference: "chicken_bar.b.q"`,
+}, {
+	input:    "foo-pkg_dashed-slice-name",
+	expected: setup.SliceKey{Package: "foo-pkg", Slice: "dashed-slice-name"},
+}, {
+	input: "foo-_-bar",
+	err:   `invalid slice reference: "foo-_-bar"`,
+}, {
+	input: "foo_bar-",
+	err:   `invalid slice reference: "foo_bar-"`,
+}, {
+	input: "foo-_bar",
+	err:   `invalid slice reference: "foo-_bar"`,
+}, {
+	input: "-foo_bar",
+	err:   `invalid slice reference: "-foo_bar"`,
+}, {
+	input:    "foo+_bar",
+	expected: setup.SliceKey{Package: "foo+", Slice: "bar"},
+}, {
+	input:    "foo_slice123",
+	expected: setup.SliceKey{Package: "foo", Slice: "slice123"},
+}}
```

Comment:
Can you add a test with two underscores "_"? Can you also add tests for several dots, several pluses, etc.? Especially corner cases with only one alphanumeric character. Ideally the tests should cover all of those to know that they are accepted.

Reply:
Added in https://github.com/canonical/chisel/pull/120/commits/be332e0d2dbd897836e1e3c6f1d1d08109701538.

```go
{
	input: "+++_bar",
	err:   `invalid slice reference: .*`,
}, {
	input: "..._bar",
	err:   `invalid slice reference: .*`,
}
```
Note that the `err` string contains `.*` at the end to not explicitly write the name of the slice in test data because the `+` and `.` values are not interpreted literally in regex (used in `ErrorMatches`).

Reply:
Right, what about escaping the `.` instead? 
```go
}, {
	input: "..._bar",
	err:   `invalid slice reference: "\.\.\._bar"`,
}
```

and one last nitpick, can you group the tests that fail and the ones that don't? So first the successful ones and then the ones that fail.

Reply:
Both done in https://github.com/canonical/chisel/pull/120/commits/1f8cedfed84ce595093e6f95d962bbac37918432.